omod/src/test/java/org/openmrs/module/oauth2login/web/JwtUtilsTest.java

```diff
@@ -0,0 +1,194 @@
+import io.jsonwebtoken.impl.DefaultClaims;
+import io.jsonwebtoken.lang.DateFormats;
+import io.jsonwebtoken.security.Keys;
+import io.jsonwebtoken.security.SignatureException;
+
+@RunWith(PowerMockRunner.class)
+@PrepareForTest({ OpenmrsUtil.class, FileUtils.class, Utils.class, HttpUtils.class })
+public class JwtUtilsTest {
+	
+	@Mock
+	private Properties mockProps;
+	
+	@Mock
+	private File mockKeyFile;
+	
+	@Rule
+	public ExpectedException ee = ExpectedException.none();
+	
+	@Before
+	public void setup() {
+		MockitoAnnotations.initMocks(this);
+		mockStatic(OpenmrsUtil.class);
+		mockStatic(Utils.class);
+		mockStatic(FileUtils.class);
+		mockStatic(HttpUtils.class);
+		Whitebox.setInternalState(JwtUtils.class, PublicKey.class, (Object) null);
+		Whitebox.setInternalState(JwtUtils.class, JsonWebKeySet.class, (Object) null);
+		Whitebox.setInternalState(JwtUtils.class, "keysInitialized", false);
+	}
+	
+	@Test
+	public void getPublicKey_shouldLookUpTheKeyFromAnOauthProperty() throws Exception {
+		final String key = IOUtils.toString(getClass().getClassLoader().getResourceAsStream("publicKey.txt"), "UTF-8");
+		PublicKey expectedKey = JwtUtils.stringToPublicKey(key.trim());
+		when(mockProps.getProperty(OAUTH_PROP_KEY)).thenReturn(key);
+		assertEquals(expectedKey, JwtUtils.getPublicKey(null, mockProps));
+	}
+	
+	@Test
+	public void getPublicKey_shouldLookUpTheKeyFromAFile() throws Exception {
+		final String key = IOUtils.toString(getClass().getClassLoader().getResourceAsStream("publicKey.txt"), "UTF-8");
+		final String testFile = "test.txt";
+		when(mockProps.getProperty(OAUTH_PROP_KEY_FILE)).thenReturn(testFile);
+		when(Utils.getFileInAppDataDirectory(testFile)).thenReturn(mockKeyFile);
+		when(mockKeyFile.exists()).thenReturn(true);
+		when(FileUtils.readFileToString(mockKeyFile, StandardCharsets.UTF_8)).thenReturn(key);
+		PublicKey expectedKey = JwtUtils.stringToPublicKey(key.trim());
+		assertEquals(expectedKey, JwtUtils.getPublicKey(null, mockProps));
+	}
+	
+	@Test
+	public void getPublicKey_shouldNotLookUpTheKeyFromTheIdentityProviderIfNoUrlIsSet() throws Exception {
+		Assert.assertNull(JwtUtils.getPublicKey(null, mockProps));
+		PowerMockito.verifyStatic(never());
+		HttpUtils.getJsonWebKeys(anyString());
+	}
+	
+	@Test
+	public void getPublicKey_shouldLookUpTheKeyFromTheIdentityProviderIfUrlIsSet() throws Exception {
+		final String url = "http://someurl.com";
+		when(mockProps.getProperty(OAUTH_PROP_KEYS_URL)).thenReturn(url);
+		final String keysJson = IOUtils.toString(getClass().getClassLoader().getResourceAsStream("keys.json"), "UTF-8");
+		JsonWebKeySet jsonWebKeySet = new JsonWebKeySet(keysJson);
+		when(HttpUtils.getJsonWebKeys(url)).thenReturn(keysJson);
+		Key expectedKey = jsonWebKeySet.getJsonWebKeys().get(0).getKey();
+		final String key = IOUtils.toString(getClass().getClassLoader().getResourceAsStream("jwtToken.txt"), "UTF-8");
+		assertEquals(expectedKey, JwtUtils.getPublicKey(key, mockProps));
+	}
```

Comment:
Like you did for `OAuth2ServiceAccountFilterTest`, could you make it clear visually where is the setup section, where is the replay section and where is the verify section? Especially for bigger tests.

I literally comment these parts usually eg: https://github.com/openmrs/openmrs-module-attachments/blob/bd8f77ba063203a6898b33ef4a856c64444b9738/omod-1.10/src/test/java/org/openmrs/module/attachments/rest/AttachmentRestController1_10Test.java#L156-L169

Reply:
Some code is self explaining, adding extra comments might be unnecessary noise especially for tests e.g adding comments to a no arg constructor but if that's how you prefer it I can do that